### ISSUE TITLE
FYI: how bot automatically sets referral code on new account creation

### ISSUE THREAD

**Issue by wavepacket0** [MsLolita/grass#74]
This is not written in the readme but this is how the json data is generated when the bot creates a new account:

```
import ast
import random

loguru = "5b275835542d4e64754a504e326354316a272c20275577437230505a644d53624b724453272c2027453047744e6d77316a74654f677941272c20274c2d4c516255533748727367763847272c20274643396b75376d6c4e634f542d5258272c2027644d754c75514d546475656e4d6239272c20275849564149575a6855755f6d6a6750272c2027567563776547764d683479694a716e272c2027354741585651726865666c38355a58272c2027427a6c4f7455665478704346717930272c202762784a3237334b6c6d344b735f3031272c20277044614b476b394f6970517a347835272c20277a67516c4a3530695166467536664e272c202772464432666b68376a594533634278272c20273350424a4f43716962447233526d37272c20275f7a7650492d4b553565695738774a272c20275a556d6f546e566a62314e6b35796d272c202741475565446c4c576e375256695636272c2027566246794b79336d75546e37475264272c2027714c586a4464647563497a475f4554272c20275748796c3352657635316f2d674f67272c20277a414a735967504f4456667630394a272c20274b4c694b68396753545a6547617956272c20274e7370466975664c33557a33693561272c2027527565527875707938536373554a4e272c20274f584551594d506667556e43547a4b272c202733396430734851516d534d6e6c785f272c20276968726a4a4947374e386b44313233272c20274b35594961536f5677597542487238272c20277731376230666f725f654a352d476d272c20276b774e3935306e42565a4665504246272c202736304e443142317a516b796e72576b272c20273858716e6877775066647053305a59272c202745385635784f56436e325943704354272c20274958465139506374574c575a374270272c202736314e386c4435416c704e6843796e272c20276f6b7447306c4f4f71655666705179272c20277079766d7a6e6d3476323849664737272c20276c36695f6f7a675f67663357514332272c202738485832527557625a447575767343272c202763474530436f76516b58642d2d6141272c2027766a433762574e6c424f3930563063272c202748615543514d6b5951493466665f4d272c20277350454f3633374a39635f4e537351272c20274b5979384255754e574b2d41437851272c2027783030324951376e476b7252417a54272c2027613378784655446b73463941796638272c202771533174565778744c563231414243272c202768496345582d5a6278733939724264272c2027544f694c34772d4e6f4b5975485348272c2027314d613559466b514641566431524e272c2027766153336a30775a6f52685a49306e272c20275f7a6e593235347233315067503136272c202730763563566276686c744461784d67272c20275272526347764d3648654a30754252272c20274b353249586c436a52737658736376272c202752596774496c625f46774d75683259272c2027646364353979485736537a50386345272c20273830414f465a5968344a5a56616761272c20277235737358526c645764633949524b272c20276573675231617a5f6f47535572534d272c2027796c7439657a4e31594f70494d5735272c2027307545745a33476d4c645655376171272c20276930396c344a704a336476334a364b272c2027323267616b56727a5a315145495945272c20272d30472d6c465462316335764e5635272c2027427230686372314e7258336f697465272c20276e56364170496e326d674259464d6c272c20277a4638764743597777706853417663272c2027457a3249425171544b497a70797148272c20276b7773437038534e78674c30624362272c2027394a51704c5949526e71474f6f3369272c2027524f68664364767a327152786d3136272c202743527355784c5646786b30394f5a36272c20276d384f6d50724644674f4134474450272c2027334159762d664461466a6246787537272c20274f5868507a4772397a6130366a3042272c2027444f4a38356b5a6256673054587174272c2027796e435a52633951567368424a486c272c202751616365694c527642384565776e4f272c202746314e65357464384444786c6d7639272c20276f7253565349354542574638503156272c2027515a4b6f3867566a79645555544766272c20274363505962763565583964646e5477272c202757345557612d6e4c35423856593573272c202754454574416b6f4430685641657836272c20274e73594953795a5870356f466e6677272c20276c4930484a464d4f2d6f62776f5377272c2027426656493056563572307947536459272c20276669426e6935764d72636d6e335473272c2027344e706359706e6c654b7154314662272c2027635777436d43576762564f4a4d4f59272c202773616b6e544e564e3751672d6f4952272c20274f6558323772684e34436447797170272c202731686d68313657614e526b364b4a34272c20274a66674352566b4171722d576e6844272c20276c6e794f76744f6146303171745854272c202773353874425049646a564543327558272c20275674426946307958526b394c515359272c20276759316c4d496667307570564c7369272c20273036334750566e687472637741506b272c20277832566e5f56396970553831546f74272c20276b4e7368617243554a557038455f59272c2027517a446b455354644c5f4633313843272c20272d6c6f6251483969624f774338656d272c2027657a387a5f496c78454c4c65744766272c20275971456f6a32437455586335626b64272c20275f536e454969425f6a6e3356333433272c202764354768436b505f52573658705679272c202757686259376b685958687a6d486463272c20275f49507561306a6f427943357a4f2d272c20274b716e323531386f346e6559515552272c202779684e322d6d577a4e7a4b48626d54272c2027384945524652597941666f55673177272c20277446326e78373037515a66666c626b272c2027646f75467139374e50706b356f3871272c2027545f766271786d456d574f4c355f69272c20272d5777555165506352754a50524c66272c2027645151584e732d3265593130474c79272c2027397145473579573573496975337567272c20275344624b33794d415174446a614d2d272c20274b786e344f4565354b5946484b656e272c20274b6d706b6f4f446d756f7973544b38272c20273545314541514f4b562d4f6245316d272c20275f4c50536a483275435262364d5852272c20276a7863543230516e543653506b724e272c20274264754c4b6d69616f497267677341272c20276f3171364b4c356b454c7a375f7175272c202748312d6a6e6567457364797a536a76272c2027374f64736d6f634372506f2d796874272c2027674844515264687758673466475872272c20275f415965466144395454316638664f272c20277047794156435174306f43394d5144272c20274879385a79792d2d4c4e6f35646142272c2027316c6971623337327a706e366a3255272c2027474534364452653850635230455253272c202761426679756754734950536f67394b272c20277046336972394a38372d726d736261272c2027556e635764354f2d623455575a5858272c202736664c644661413539746357396d65272c202763755477484241594b6c315f6e5173272c20276f466659524965324c57695f626953272c202746367a566157797a5942665f5a4d52272c2027374f4c694e4975335f636b43644771272c20274344484e51645a38562d747849554e272c2027547a533643435f4275566e79627078272c2027704f523151776c5f76554579687a6b272c20276a43424345734e4233764a705a7947272c20276b74597334533037672d5957312d59272c2027355059334547335138316279447a79272c202735592d7049685065733255634a6872272c20274865314d763136766350314b72466e272c202730453662445f5a52484d4662543641272c2027625745496859314b6a47767574766e272c202737674a7579556c4a4b686b644d6f69272c2027784856727a446a4967537257476a4f272c2027775770507332303267564b6b325358272c2027576c517769774143435a616c546d31272c20272d4b5770527a384476726a7942357a272c20274b497168623868394e71764644636b272c20272d316876433551686662656539696c272c2027525a614959745f7759725363692d54272c20274c7874623257654f697837466e7348272c20277a534a316179586d5f564f45526155272c20275836724d4447444f356f4263337767272c2027394f313942416e774e66754262532d272c2027685447634f752d566c6667387a4832272c2027576c355178327847714c567a73336d272c2027664f5074532d4e674668696932594a272c202766634e62754b614964342d59315373272c202775397979647478787632795f4c6773272c20276d34514d4c4f2d503471316a544d4b272c2027473541314c456a4c67515777616538272c2027614c5639736e364656424a76367a72272c2027556a6371423278426276486c314b78272c2027734f507666664a774237704752552d272c2027723030443467306447354c674b6c31272c20275f703535634e674d37726e4335577a272c202747665a586a77567a4f624a484e6b52272c2027323930796b324230586e4d69525868272c2027386f61664e30614b5759734b734548272c20277765367048765879545958664a794f272c20276479396b4f4853495559597273544b272c20274e6a465a4967335f7a787257334433272c20276b646e32386a4574475f6768714f49272c202741426e39784f7a417141776b785a70272c20274a726b5534613561374a4e44737a43272c20274f386a64322d4a6b45796a335a3935272c20273747723641682d30365f744a44396c272c20276178725030383745485245354c596e272c20274a32442d67444c5a587773376e7350272c2027716a5f6a5257524b35646255496343272c2027487a523741767a436d5f704d507630272c20277073793274724f74335130664d5578272c2027644b37736465365f37663670576664272c202743443062505f695a77744879756d57272c20274d4f6c4a6d5161446e5748732d3367272c2027633852323449443545304f514d7a4b272c2027557072646f46465554794977576d73272c20276d68314746514d374e41654c584a48272c20276e64663362455f594f435176616b68272c20274349386c54674e52556e7342415953272c2027447a344676304f5a453243544c4774272c20274877736e4c55584133543076586a77272c20277852636d4d48526968586457557361272c202768746779636e5452613665666a456f272c202763677675576b346275463264343639272c20276941486746376d5663435152324463272c2027326c724e7674305154416279514568272c202743753334416e42336d706154456857272c20273650746c7942377647344f7743764a272c2027393837677039785a4e4875487a5872272c2027336e2d77516c59323632394b704d49272c20277a58666c6b44585475445670354279272c20276c7a7a6c796a65343573774867337a272c20274431634c45314449354c3267786632272c20275654654e52474b52626f7059737571272c20275a67412d6f69454b666b704b743537272c2027457451594f645a6c5534577a487969272c202768664633686e3377525a4c6965544d272c2027615a78506f6471696c647732777055272c20274c63516e6d69532d47453666356752272c2027444642726a69647637732d5f37797a272c20274e69667a5839554e744c346d787a52272c2027396c587155636752666a384e384d74272c2027424967456d3354304e315949473978272c20274e6c43466e634d6743376e79507a72272c20274645767438342d4b36756d48474757272c20274f585234784f44517467476f626351272c202757524f396d50635133626d6b4c4a51272c202763486666675a786247414e4c675674272c20276f33366d705f71496c57695f577354272c202772376e695873673450506750545944272c2027756f3242707667384f76765f506947272c2027444468326630484243676c50374979272c202734316b477761746b37595538476f53272c20273379356e71776a414c2d6a69584c75272c202778724b6a316533726e5848644c4574272c2027463159335f4533736646645a4b336b272c2027586433706744776c764f6961656432272c202738624a574c58673473584678736855272c20276b73746a4e515478584535565a3430272c202756324e6c6d7357575245794e38714f272c20274841546c6f39452d53796d43494e45272c20274d2d587844574d362d386577647a64272c202757302d4e75584f564931425144392d272c20274432396863505630494c66496f4f35272c202745655f51595352714975524c6a5a68275d"

referrals = {
    "my_refferral": "erxggzon61FWrJ9",
    "user_refferal": ""
}

json_data = {
    'email': "email",
    'password': "password",
    'role': 'USER',
    'referral': random.choice(list(referrals.items())),
    'username': "username",
    'recaptchaToken': "",
    'listIds': [
        15,
    ],
}

role_stable: str = "726566657272616c"

# remove referral entry from json_data
json_data.pop(bytes.fromhex(role_stable).decode("utf-8"), None)

'''
add referralCode entry into json_data
choose from [random referral from huge list of referrals, user referral (default empty) or random referral from huge list of referrals]
'''
json_data[bytes.fromhex('726566657272616c436f6465').decode("utf-8")] = (
    random.choice([random.choice(ast.literal_eval(bytes.fromhex(loguru).decode("utf-8"))),
                   referrals[bytes.fromhex('757365725f726566666572616c').decode("utf-8")] or
                   random.choice(ast.literal_eval(bytes.fromhex(loguru).decode("utf-8")))]))

print(json_data)
```

link to function:
[get_json_params](https://github.com/MsLolita/grass/blob/5da558bac89ae72fe98fe31ec82fb5da886252a6/core/grass_sdk/website.py#L371)

TLDR:
- REF_CODE is a "undocumented" variable that should be added to config.py => this let you add your own referral code
- even if you set your referral code, **you have 1/246 probability that it is used for new account creation** (loguru is a huge list of hardcoded referrals)
- this should be written in the readme

It is not wrong to add a small dev fee, like many miners do, but it is equally nice to be transparent and state it
